README.md
# rfc2047-decoder [![Crates.io](https://img.shields.io/crates/v/rfc2047-decoder?style=flat-square)](https://crates.io/crates/rfc2047-decoder) [![Crates.io](https://img.shields.io/crates/d/rfc2047-decoder?style=flat-square)](https://crates.io/crates/rfc2047-decoder)

# State
This project is considered as finished, only bugs will be fixed so don't wonder, if the last commit is
a long time ago.

# Introduction

Rust library for decoding [RFC 2047 MIME Message Headers](https://tools.ietf.org/html/rfc2047).

```rust
use rfc2047_decoder;

fn main() {
    let encoded_str = "=?UTF-8?Q?str?=";
    let decoded_str = "str";

    // using the decode helper (default options)
    assert_eq!(
        rfc2047_decoder::decode(encoded_str.as_bytes()).unwrap(),
        decoded_str
    );

    // using the decoder builder (custom options)
    assert_eq!(
        rfc2047_decoder::Decoder::new()
            .too_long_encoded_word_strategy(rfc2047_decoder::RecoverStrategy::Skip)
            .decode(encoded_str.as_bytes())
            .unwrap(),
        decoded_str
    );
}
```

## Sponsoring

[![github](https://img.shields.io/badge/-GitHub%20Sponsors-fafbfc?logo=GitHub%20Sponsors&style=flat-square)](https://github.com/sponsors/soywod)
[![paypal](https://img.shields.io/badge/-PayPal-0079c1?logo=PayPal&logoColor=ffffff&style=flat-square)](https://www.paypal.com/paypalme/soywod)
[![ko-fi](https://img.shields.io/badge/-Ko--fi-ff5e5a?logo=Ko-fi&logoColor=ffffff&style=flat-square)](https://ko-fi.com/soywod)
[![buy-me-a-coffee](https://img.shields.io/badge/-Buy%20Me%20a%20Coffee-ffdd00?logo=Buy%20Me%20A%20Coffee&logoColor=000000&style=flat-square)](https://www.buymeacoffee.com/soywod)
[![liberapay](https://img.shields.io/badge/-Liberapay-f6c915?logo=Liberapay&logoColor=222222&style=flat-square)](https://liberapay.com/soywod)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.3] - 2023-12-23

- Fix versioning

## [1.0.2] - 2023-12-23

- Bumping dependencies

## [1.0.1] - 2023-09-17
- extending docs about some structs
- fixing `Display` implementation of `TooLongEncodedWord`

## [1.0.0] - 2023-09-16

### Changed

- `lexer::Error` gets the alias `LexerError`
- `parser::Error` gets the alias `ParserError`
- `evaluator::Error` gets the alias `EvaluatorError`
- `Decoder.too_long_encoded_word` has a strategy instead of a `bool`

## [0.2.2] - 2023-03-29

### Changed

- Bumped `base64` to `v0.21.0`.
- Bumped `chumsky` to `v0.9.2`.

## [0.2.1] - 2023-01-08

### Fixed

- Fixed discarded errors [#20].

## [0.2.0] - 2022-10-11

### Added

- Added Nix support
- Allowed decoding strings bigger than 76 chars [#15]

### Changed

- Renamed error variants to match the [Rust API
  guidelines](https://rust-lang.github.io/api-guidelines/naming.html#names-use-a-consistent-word-order-c-word-order):
  - `lexer::Error::EncodingIssue` becomes `ParseBytesError`
  - `lexer::Error::EncodedWordTooLong` becomes
    `ParseEncodedWordTooLongError`
  - `parser::Error::UnknownCharset` becomes `ParseEncodingError`
  - `parser::Error::UnknownCharset` has been removed (unused)
  - `parser::Error::UnknownEncoding` becomes `ParseEncodingError`
  - `parser::Error::EncodedWordTooBig` becomes
    `ParseEncodingTooBigError`
  - `parser::Error::EmptyEncoding` becomes `ParseEncodingEmptyError`
  - `evaluator::Error::DecodeUtf8` becomes `DecodeUtf8Error`
  - `evaluator::Error::DecodeBase64` becomes `DecodeBase64Error`
  - `evaluator::Error::DecodeQuotedPrintable` becomes
    `DecodeQuotedPrintableError`

## [0.1.3] - 2022-10-10

### Fixed

- Max length of encoded words [#1]
- Manage tokens special chars [#3]

### Changed

- Refactored parser using chumsky [#7]

## [0.1.2] - 2020-12-30

### Fixed

- Multiple encoded words separator

## [0.1.1] - 2020-12-30

### Added

- Added evaluator with AST

### Changed

- Decoded fn accepts now `&[u8]` instead of `&str`

### Fixed

- Removed space between encoded words [#2]

## [0.1.0] - 2020-12-28

First official release.

[unreleased]: https://github.com/soywod/rfc2047-decoder/compare/v0.2.2...HEAD
[0.2.2]: https://github.com/soywod/rfc2047-decoder/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/soywod/rfc2047-decoder/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/soywod/rfc2047-decoder/compare/v0.1.3...v0.2.0
[0.1.3]: https://github.com/soywod/rfc2047-decoder/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/soywod/rfc2047-decoder/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/soywod/rfc2047-decoder/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/soywod/rfc2047-decoder/releases/tag/v0.1.0

[#1]: https://github.com/soywod/rfc2047-decoder/issues/1
[#2]: https://github.com/soywod/rfc2047-decoder/issues/2
[#3]: https://github.com/soywod/rfc2047-decoder/issues/3
[#7]: https://github.com/soywod/rfc2047-decoder/issues/7
[#15]: https://github.com/soywod/rfc2047-decoder/issues/15
[#20]: https://github.com/soywod/rfc2047-decoder/issues/20


src/lib.rs
//! # Introduction
//! This crate provides a [Decoder] and the function [decode], in order to decode
//! encoded words as described in the [RFC 2047].
//!
//! [RFC 2047]: https://datatracker.ietf.org/doc/html/rfc2047
//!
//! # Where to start looking
//! You will likely want to start looking into [Decoder] and/or the [decode]
//! to use this crate.

mod decoder;
pub use decoder::{Decoder, Error, RecoverStrategy};

mod evaluator;
mod lexer;
mod parser;

pub use evaluator::Error as EvaluatorError;
pub use lexer::{Error as LexerError, TooLongEncodedWords};
pub use parser::Error as ParserError;

/// Decodes the given RFC 2047 MIME Message Header encoded string
/// using a default decoder.
///
/// This function equals doing `Decoder::new().decode`.
///
/// # Example
/// ```
/// use rfc2047_decoder::{decode, Decoder};
///
/// let encoded_message = "=?ISO-8859-1?Q?hello_there?=".as_bytes();
/// let decoded_message = "hello there";
///
/// // This ...
/// assert_eq!(decode(encoded_message).unwrap(), decoded_message);
///
/// // ... equals this:
/// assert_eq!(Decoder::new().decode(encoded_message).unwrap(), decoded_message);
/// ```
pub fn decode<T: AsRef<[u8]>>(encoded_str: T) -> Result<String, Error> {
    Decoder::new().decode(encoded_str)
}


src/lexer.rs
use chumsky::{prelude::Simple, text::whitespace, Parser};
use std::{collections::HashSet, result, fmt::Display};
use thiserror::Error;

use crate::{decoder::RecoverStrategy, Decoder};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct TooLongEncodedWords(Vec<String>);

impl TooLongEncodedWords {
    pub fn new(encoded_words: Vec<String>) -> Self {
        Self(encoded_words)
    }
}

impl Display for TooLongEncodedWords {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut message = String::new();

        if !self.0.is_empty() {
            message = self.0[0].clone();

            for encoded_word in self.0.iter().skip(1) {
                message.push_str(&format!(", {}", encoded_word));
            }
        }

        f.write_str(&message)
    }
}

#[derive(Error, Debug, Clone, PartialEq)]
pub enum Error {
    #[error("cannot parse bytes into tokens")]
    ParseBytesError(Vec<Simple<u8>>),
    #[error("Cannot parse the following encoded words, because they are too long: {0}")]
    ParseEncodedWordTooLongError(TooLongEncodedWords),
}

type Result<T> = result::Result<T, Error>;

const QUESTION_MARK: u8 = b'?';
const SPACE: u8 = b' ';
const AMOUNT_DELIMITERS: usize = "=????=".len();

pub type Tokens = Vec<Token>;

#[derive(Debug, Clone, PartialEq, Hash, Eq)]
pub enum Token {
    ClearText(Vec<u8>),
    EncodedWord {
        charset: Vec<u8>,
        encoding: Vec<u8>,
        encoded_text: Vec<u8>,
    },
}

impl Token {
    pub const ENCODED_WORD_PREFIX: &'static [u8] = "=?".as_bytes();
    pub const ENCODED_WORD_SUFFIX: &'static [u8] = "?=".as_bytes();
    pub const MAX_ENCODED_WORD_LENGTH: usize = 75;

    /// Returns the length of the encoded word including the delimiters
    pub fn len(&self) -> usize {
        match self {
            Self::ClearText(_) => 1,
            Self::EncodedWord {
                charset,
                encoding,
                encoded_text,
            } => charset.len() + encoding.len() + encoded_text.len() + AMOUNT_DELIMITERS,
        }
    }

    pub fn is_encoded_word(&self) -> bool {
        match self {
            Self::EncodedWord {..} => true,
            _ => false,
        }
    }

    pub fn get_bytes(&self) -> Vec<u8> {
        match self {
            Token::ClearText(token) => (*token).clone(),
            Token::EncodedWord {
                charset,
                encoding,
                encoded_text,
            } => {
                let mut bytes = Vec::new();
                bytes.extend(charset);
                bytes.extend(encoding);
                bytes.extend(encoded_text);
                bytes
            }
        }
    }

    pub fn get_encoded_word(
        ((charset, encoding), encoded_text): ((Vec<u8>, Vec<u8>), Vec<u8>),
    ) -> Self {
        Self::EncodedWord {
            charset,
            encoding,
            encoded_text,
        }
    }
}

pub fn run(encoded_bytes: &[u8], decoder: Decoder) -> Result<Tokens> {
    let tokens = get_parser(&decoder)
        .parse(encoded_bytes)
        .map_err(|err| Error::ParseBytesError(err))?;

    validate_tokens(tokens, &decoder)
}

fn get_parser(decoder: &Decoder) -> impl Parser<u8, Tokens, Error = Simple<u8>> {
    use chumsky::prelude::*;

    let encoded_words_in_a_row = {
        let following_encoded_word =
            whitespace().ignore_then(encoded_word_parser(&decoder).rewind());
        encoded_word_parser(&decoder).then_ignore(following_encoded_word)
    };

    let single_encoded_word = encoded_word_parser(&decoder);
    let single_clear_text = clear_text_parser(&decoder);

    encoded_words_in_a_row
        .or(single_encoded_word)
        .or(single_clear_text)
        .repeated()
}

fn clear_text_parser(decoder: &Decoder) -> impl Parser<u8, Token, Error = Simple<u8>> {
    use chumsky::prelude::*;

    const DEFAULT_EMPTY_INPUT_ERROR_MESSAGE: &str = "got empty input";

    take_until(encoded_word_parser(&decoder).rewind().ignored().or(end())).try_map(
        |(chars, ()), span| {
            if chars.is_empty() {
                Err(Simple::custom(span, DEFAULT_EMPTY_INPUT_ERROR_MESSAGE))
            } else {
                Ok(Token::ClearText(chars))
            }
        },
    )
}

fn encoded_word_parser(decoder: &Decoder) -> impl Parser<u8, Token, Error = Simple<u8>> {
    use chumsky::prelude::*;

    let skip_encoded_word_length = decoder.too_long_encoded_word;

    let check_encoded_word_length = move |token: Token, _span| {
        if token.len() > Token::MAX_ENCODED_WORD_LENGTH {
            match skip_encoded_word_length {
                RecoverStrategy::Skip => Ok(Token::ClearText(token.get_bytes())),
                // Hint: The lexer 
                RecoverStrategy::Decode | RecoverStrategy::Abort => Ok(token),
            }
        } else {
            Ok(token)
        }
    };

    let is_especial = |c: u8| get_especials().contains(&c);

    let token = filter(move |&c: &u8| c != SPACE && !c.is_ascii_control() && !is_especial(c));
    let charset = token.repeated().at_least(1).collect::<Vec<u8>>();
    let encoding = token.repeated().at_least(1).collect::<Vec<u8>>();
    let encoded_text = filter(|&c: &u8| c != QUESTION_MARK && c != SPACE)
        .repeated()
        .collect::<Vec<u8>>();

    just(Token::ENCODED_WORD_PREFIX)
        .ignore_then(charset)
        .then_ignore(just(QUESTION_MARK))
        .then(encoding)
        .then_ignore(just(QUESTION_MARK))
        .then(encoded_text)
        .then_ignore(just(Token::ENCODED_WORD_SUFFIX))
        .map(Token::get_encoded_word)
        .try_map(check_encoded_word_length)
}

fn get_especials() -> HashSet<u8> {
    "()<>@,;:/[]?.=".bytes().collect()
}

fn validate_tokens(tokens: Tokens, decoder: &Decoder) -> Result<Tokens> {
    if let Some(too_long_encoded_words) = get_too_long_encoded_words(&tokens, decoder) {
        return Err(Error::ParseEncodedWordTooLongError(too_long_encoded_words));
    }

    Ok(tokens)
}

fn get_too_long_encoded_words(tokens: &Tokens, decoder: &Decoder) -> Option<TooLongEncodedWords> {
    let strategy = decoder.too_long_encoded_word;
    let mut too_long_encoded_words: Vec<String> = Vec::new();

    for token in tokens.into_iter() {
        if token.is_encoded_word() && token.len() > Token::MAX_ENCODED_WORD_LENGTH && strategy == RecoverStrategy::Abort {
            let encoded_word = String::from_utf8(token.get_bytes()).unwrap();
            too_long_encoded_words.push(encoded_word);
        }
    }

    if too_long_encoded_words.is_empty() {
        None
    } else {
        Some(TooLongEncodedWords::new(too_long_encoded_words))
    }
}

#[cfg(test)]
mod tests {
    use crate::{lexer::Token, Decoder};

    use super::get_parser;
    use chumsky::Parser;

    #[test]
    fn test_encoded_word() {
        let parser = get_parser(&Decoder::new());
        let message = "=?ISO-8859-1?Q?Yeet?=".as_bytes();

        let parsed = parser.parse(message).unwrap();

        assert_eq!(
            parsed,
            vec![Token::EncodedWord {
                charset: "ISO-8859-1".as_bytes().to_vec(),
                encoding: "Q".as_bytes().to_vec(),
                encoded_text: "Yeet".as_bytes().to_vec(),
            }]
        );
    }

    #[test]
    fn test_clear_text() {
        let parser = get_parser(&Decoder::new());
        let message = "I use Arch by the way".as_bytes();

        let parsed = parser.parse(message).unwrap();

        assert_eq!(
            parsed,
            vec![Token::ClearText(
                "I use Arch by the way".as_bytes().to_vec()
            )]
        );
    }

    // The following examples are from the encoded-form table in section 8:
    // https://datatracker.ietf.org/doc/html/rfc2047#section-8
    #[test]
    fn test_encoded_from_1() {
        let parser = get_parser(&Decoder::new());
        let message = "=?ISO-8859-1?Q?a?=".as_bytes();

        let parsed = parser.parse(message).unwrap();

        assert_eq!(
            parsed,
            vec![Token::EncodedWord {
                charset: "ISO-8859-1".as_bytes().to_vec(),
                encoding: "Q".as_bytes().to_vec(),
                encoded_text: "a".as_bytes().to_vec()
            }]
        );
    }

    // see test_encoded_from_1
    #[test]
    fn test_encoded_from_2() {
        let parser = get_parser(&Decoder::new());
        let message = "=?ISO-8859-1?Q?a?= b".as_bytes();

        let parsed = parser.parse(message).unwrap();

        assert_eq!(
            parsed,
            vec![
                Token::EncodedWord {
                    charset: "ISO-8859-1".as_bytes().to_vec(),
                    encoding: "Q".as_bytes().to_vec(),
                    encoded_text: "a".as_bytes().to_vec(),
                },
                Token::ClearText(" b".as_bytes().to_vec()),
            ]
        );
    }

    // see test_encoded_from_1
    #[test]
    fn test_encoded_from_3() {
        let parser = get_parser(&Decoder::new());
        let message = "=?ISO-8859-1?Q?a?= =?ISO-8859-1?Q?b?=".as_bytes();

        let parsed = parser.parse(message).unwrap();

        assert_eq!(
            parsed,
            vec![
                Token::EncodedWord {
                    charset: "ISO-8859-1".as_bytes().to_vec(),
                    encoding: "Q".as_bytes().to_vec(),
                    encoded_text: "a".as_bytes().to_vec(),
                },
                Token::EncodedWord {
                    charset: "ISO-8859-1".as_bytes().to_vec(),
                    encoding: "Q".as_bytes().to_vec(),
                    encoded_text: "b".as_bytes().to_vec()
                }
            ]
        );
    }

    /// Test if parser can parse multiple encoded words in a row
    /// See: https://datatracker.ietf.org/doc/html/rfc2047#section-8
    #[test]
    fn test_multiple_encoded_words() {
        let parser = get_parser(&Decoder::new());
        let message = "=?ISO-8859-1?Q?a?= =?ISO-8859-1?Q?b?= =?ISO-8859-1?Q?c?=".as_bytes();

        let parsed = parser.parse(message).unwrap();

        assert_eq!(
            parsed,
            vec![
                Token::EncodedWord {
                    charset: "ISO-8859-1".as_bytes().to_vec(),
                    encoding: "Q".as_bytes().to_vec(),
                    encoded_text: "a".as_bytes().to_vec(),
                },
                Token::EncodedWord {
                    charset: "ISO-8859-1".as_bytes().to_vec(),
                    encoding: "Q".as_bytes().to_vec(),
                    encoded_text: "b".as_bytes().to_vec()
                },
                Token::EncodedWord {
                    charset: "ISO-8859-1".as_bytes().to_vec(),
                    encoding: "Q".as_bytes().to_vec(),
                    encoded_text: "c".as_bytes().to_vec()
                }
            ]
        );
    }

    #[test]
    fn test_ignore_mutiple_spaces_between_encoded_words() {
        let parser = get_parser(&Decoder::new());
        let message =
            "=?ISO-8859-1?Q?a?=                               =?ISO-8859-1?Q?b?=".as_bytes();

        let parsed = parser.parse(message).unwrap();

        assert_eq!(
            parsed,
            vec![
                Token::EncodedWord {
                    charset: "ISO-8859-1".as_bytes().to_vec(),
                    encoding: "Q".as_bytes().to_vec(),
                    encoded_text: "a".as_bytes().to_vec(),
                },
                Token::EncodedWord {
                    charset: "ISO-8859-1".as_bytes().to_vec(),
                    encoding: "Q".as_bytes().to_vec(),
                    encoded_text: "b".as_bytes().to_vec()
                }
            ]
        );
    }

    /// An encoded word with more then 75 chars should be parsed as a normal cleartext
    #[test]
    fn test_too_long_encoded_word() {
        let parser = get_parser(&Decoder::new());
        // "=?" (2) + "ISO-8859-1" (10) + "?" (1) + "Q" (1) + "?" (1) + 'a' (60) + "?=" (2)
        // = 2 + 10 + 1 + 1 + 1 + 60 + 2
        // = 77 => too long
        let message =
            "=?ISO-8859-1?Q?aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa?="
                .as_bytes();
        let parsed = parser.parse(message).unwrap();

        assert_eq!(parsed, vec![Token::ClearText(message.to_vec())]);
    }

    #[test]
    fn test_encoded_word_has_especials() {
        let parser = get_parser(&Decoder::new());
        let message = "=?ISO-8859-1(?Q?a?=".as_bytes();
        let parsed = parser.parse(message).unwrap();

        assert_eq!(parsed, vec![Token::ClearText(message.to_vec())]);
    }
}


src/lexer/encoded_word.rs
use std::fmt::Display;

use super::QUESTION_MARK;

pub const PREFIX: &[u8] = "=?".as_bytes();
pub const SUFFIX: &[u8] = "?=".as_bytes();
pub const MAX_LENGTH: usize = 75;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct EncodedWord {
    pub charset: Vec<u8>,
    pub encoding: Vec<u8>,
    pub encoded_text: Vec<u8>,
}

impl EncodedWord {
    pub fn new(charset: Vec<u8>, encoding: Vec<u8>, encoded_text: Vec<u8>) -> Self {
        Self {
            charset,
            encoding,
            encoded_text,
        }
    }

    pub fn from_parser(((charset, encoding), encoded_text): ((Vec<u8>, Vec<u8>), Vec<u8>)) -> Self {
        Self::new(charset, encoding, encoded_text)
    }

    /// Returns the amount of `char`s for this encoded word
    pub fn len(&self) -> usize {
        self.get_bytes(true).len()
    }

    pub fn get_bytes(&self, with_delimiters: bool) -> Vec<u8> {
        let mut bytes = Vec::new();

        if with_delimiters {
            bytes.extend(PREFIX);
            bytes.extend(&self.charset);
            bytes.extend(&[QUESTION_MARK]);
            bytes.extend(&self.encoding);
            bytes.extend(&[QUESTION_MARK]);
            bytes.extend(&self.encoded_text);
            bytes.extend(SUFFIX);
        } else {
            bytes.extend(&self.charset);
            bytes.extend(&self.encoding);
            bytes.extend(&self.encoded_text);
        }

        bytes
    }
}

impl Display for EncodedWord {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let charset = String::from_utf8(self.charset.clone()).unwrap();
        let encoding = String::from_utf8(self.encoding.clone()).unwrap();
        let encoded_text = String::from_utf8(self.encoded_text.clone()).unwrap();

        write!(f, "=?{}?{}?{}?=", charset, encoding, encoded_text)
    }
}


src/decoder.rs
use std::result;
use thiserror::Error;

use crate::{evaluator, lexer, parser};

/// The possible errors which can occur while parsing the string.
#[derive(Error, Debug, PartialEq)]
pub enum Error {
    /// Symbolises that an error occured in the lexer.
    #[error(transparent)]
    Lexer(#[from] lexer::Error),

    /// Symbolises that an error occured in the parser.
    #[error(transparent)]
    Parser(#[from] parser::Error),

    /// Symbolises that an error occured in the evaluator.
    #[error(transparent)]
    Evaluator(#[from] evaluator::Error),
}

/// Determines which strategy should be used if an encoded word isn't encoded as
/// described in the RFC.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum RecoverStrategy {
    /// Decode the encoded word although it's incorrectly encoded.
    ///
    /// # Example
    /// Take a look to [Decoder#RecoveryStrategy::Decode](Decoder#recoverstrategydecode).
    Decode,

    /// Skip the incorrectly encoded encoded word.
    ///
    /// # Example
    /// Take a look to [Decoder#RecoveryStrategy::Skip](Decoder#recoverstrategyskip).
    Skip,

    /// Abort the string-parsing and return an error.
    ///
    /// # Example
    /// Take a look to [Decoder#RecoveryStrategy::Abort](Decoder#recoverstrategyabort-default).
    Abort,
}

type Result<T> = result::Result<T, Error>;

/// Represents the decoder builder.
///
/// # Example
/// ```
/// use rfc2047_decoder::{Decoder, RecoverStrategy};
///
/// let decoder = Decoder::new()
///                 .too_long_encoded_word_strategy(RecoverStrategy::Skip);
/// let decoded_str = decoder.decode("=?UTF-8?B?c3Ry?=").unwrap();
///
/// assert_eq!(decoded_str, "str");
/// ```
#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Decoder {
    /// Determines which strategy should be used, if the parser encounters
    /// encoded words which are longer than allowed in the RFC (it's longer than 75 chars).
    pub too_long_encoded_word: RecoverStrategy,
}

impl Decoder {
    /// Equals [Decoder::default].
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the strategy if the decoder finds an encoded word which is too long.
    ///
    /// # Examples
    ///
    /// Each example uses the same encoded message:
    /// ```txt
    /// =?utf-8?B?TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZWxpdC4gVXQgaW50ZXJkdW0gcXVhbSBldSBmYWNpbGlzaXMgb3JuYXJlLg==?=
    /// ```
    /// which exceeds the maximum length of 75 chars so it's actually invalid.
    ///
    /// ## RecoverStrategy::Skip
    /// Skips the invalid encoded word and parses it as clear text.
    ///
    /// ```rust
    /// use rfc2047_decoder::{Decoder, RecoverStrategy};
    ///
    /// let message = "=?utf-8?B?TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZWxpdC4gVXQgaW50ZXJkdW0gcXVhbSBldSBmYWNpbGlzaXMgb3JuYXJlLg==?=";
    /// let decoder = Decoder::new()
    ///                 .too_long_encoded_word_strategy(RecoverStrategy::Skip);
    ///
    /// let parsed = decoder.decode(message).unwrap();
    ///
    /// // nothing changed!
    /// assert_eq!(parsed, message);
    /// ```
    ///
    /// ## RecoverStrategy::Decode
    /// Although the encoded word is invalid, keep decoding it.
    ///
    /// ```rust
    /// use rfc2047_decoder::{Decoder, RecoverStrategy};
    ///
    /// let message = "=?utf-8?B?TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZWxpdC4gVXQgaW50ZXJkdW0gcXVhbSBldSBmYWNpbGlzaXMgb3JuYXJlLg==?=";
    /// let decoder = Decoder::new()
    ///                 .too_long_encoded_word_strategy(RecoverStrategy::Decode);
    ///
    /// let parsed = decoder.decode(message).unwrap();
    ///
    /// // could you decode it? ;)
    /// let expected_result = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut interdum quam eu facilisis ornare.";
    ///
    /// assert_eq!(parsed, expected_result);
    /// ```
    ///
    /// ## RecoverStrategy::Abort (default)
    /// The parser will return an `Err` and collects all encoded words which are
    /// too long. You can use them afterwards for error messages for example.
    ///
    /// ```rust
    /// use rfc2047_decoder::{Decoder, RecoverStrategy, Error::{self, Lexer}};
    /// use rfc2047_decoder::LexerError::ParseEncodedWordTooLongError;
    /// use rfc2047_decoder::TooLongEncodedWords;
    ///
    /// let message = "=?utf-8?B?TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZWxpdC4gVXQgaW50ZXJkdW0gcXVhbSBldSBmYWNpbGlzaXMgb3JuYXJlLg==?=";
    /// // `RecoverStrategy::Abort` is the default strategy
    /// let decoder = Decoder::new();
    ///
    /// let parsed = decoder.decode(message);
    ///
    /// assert_eq!(parsed, Err(Lexer(ParseEncodedWordTooLongError(TooLongEncodedWords(vec!["=?utf-8?B?TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZWxpdC4gVXQgaW50ZXJkdW0gcXVhbSBldSBmYWNpbGlzaXMgb3JuYXJlLg==?=".to_string()])))));
    /// ```
    pub fn too_long_encoded_word_strategy(mut self, strategy: RecoverStrategy) -> Self {
        self.too_long_encoded_word = strategy;
        self
    }

    /// Decodes the given RFC 2047 MIME Message Header encoded string.
    pub fn decode<T: AsRef<[u8]>>(self, encoded_str: T) -> Result<String> {
        let text_tokens = lexer::run(encoded_str.as_ref(), self)?;
        let parsed_text = parser::run(text_tokens)?;
        let evaluated_string = evaluator::run(parsed_text)?;

        Ok(evaluated_string)
    }
}

impl Default for Decoder {
    /// Returns the decoder with the following default "settings":
    ///
    /// - `too_long_encoded_word`: [RecoverStrategy::Abort]
    fn default() -> Self {
        Self {
            too_long_encoded_word: RecoverStrategy::Abort,
        }
    }
}

#[cfg(test)]
mod tests {
    /// Here are the main-tests which are listed here:
    /// https://datatracker.ietf.org/doc/html/rfc2047#section-8
    /// Scroll down until you see the table.
    mod rfc_tests {
        use crate::decode;

        #[test]
        fn decode_encoded_word_single_char() {
            assert_eq!(decode("=?ISO-8859-1?Q?a?=").unwrap(), "a");
        }

        #[test]
        fn decode_encoded_word_separated_by_whitespace() {
            assert_eq!(decode("=?ISO-8859-1?Q?a?= b").unwrap(), "a b");
        }

        #[test]
        fn decode_two_encoded_chars() {
            assert_eq!(
                decode("=?ISO-8859-1?Q?a?= =?ISO-8859-1?Q?b?=").unwrap(),
                "ab"
            );
        }

        #[test]
        fn whitespace_between_two_encoded_words_should_be_ignored() {
            assert_eq!(
                decode("=?ISO-8859-1?Q?a?=  =?ISO-8859-1?Q?b?=").unwrap(),
                "ab"
            );
        }

        #[test]
        fn whitespace_chars_between_two_encoded_words_should_be_ignored() {
            assert_eq!(
                decode(
                    "=?ISO-8859-1?Q?a?=               
                     =?ISO-8859-1?Q?b?="
                )
                .unwrap(),
                "ab"
            );
        }

        #[test]
        fn whitespace_encoded_in_encoded_word() {
            assert_eq!(decode("=?ISO-8859-1?Q?a_b?=").unwrap(), "a b");
        }

        #[test]
        fn ignore_whitespace_between_two_encoded_words_but_not_the_encoded_whitespace() {
            assert_eq!(
                decode("=?ISO-8859-1?Q?a?= =?ISO-8859-2?Q?_b?=").unwrap(),
                "a b"
            );
        }
    }

    /// Those are some custom tests
    mod custom_tests {
        use crate::decode;

        #[test]
        fn clear_empty() {
            assert_eq!(decode("").unwrap(), "");
        }

        #[test]
        fn clear_with_spaces() {
            assert_eq!(decode("str with spaces").unwrap(), "str with spaces");
        }

        #[test]
        fn utf8_qs_empty() {
            assert_eq!(decode("").unwrap(), "");
        }

        #[test]
        fn utf8_qs_with_str() {
            assert_eq!(decode("=?UTF-8?Q?str?=").unwrap(), "str");
        }

        #[test]
        fn utf8_qs_with_spaces() {
            assert_eq!(
                decode("=?utf8?q?str_with_spaces?=").unwrap(),
                "str with spaces"
            );
        }

        #[test]
        fn utf8_qs_with_spec_chars() {
            assert_eq!(
                decode("=?utf8?q?str_with_special_=C3=A7h=C3=A0r=C3=9F?=").unwrap(),
                "str with special çhàrß"
            );
        }

        #[test]
        fn utf8_qs_double() {
            assert_eq!(
                decode("=?UTF-8?Q?str?=\r\n =?UTF-8?Q?str?=").unwrap(),
                "strstr"
            );
            assert_eq!(
                decode("=?UTF-8?Q?str?=\n =?UTF-8?Q?str?=").unwrap(),
                "strstr"
            );
            assert_eq!(decode("=?UTF-8?Q?str?= =?UTF-8?Q?str?=").unwrap(), "strstr");
            assert_eq!(decode("=?UTF-8?Q?str?==?UTF-8?Q?str?=").unwrap(), "strstr");
        }

        #[test]
        fn utf8_b64_empty() {
            assert_eq!(decode("=?UTF-8?B??=").unwrap(), "");
        }

        #[test]
        fn utf8_b64_with_str() {
            assert_eq!(decode("=?UTF-8?B?c3Ry?=").unwrap(), "str");
        }

        #[test]
        fn utf8_b64_with_spaces() {
            assert_eq!(
                decode("=?utf8?b?c3RyIHdpdGggc3BhY2Vz?=").unwrap(),
                "str with spaces"
            );
        }

        #[test]
        fn utf8_b64_with_spec_chars() {
            assert_eq!(
                decode("=?utf8?b?c3RyIHdpdGggc3BlY2lhbCDDp2jDoHLDnw==?=").unwrap(),
                "str with special çhàrß"
            );
        }

        #[test]
        fn utf8_b64_trailing_bit() {
            assert_eq!(
                decode("=?utf-8?B?UG9ydGFsZSBIYWNraW5nVGVhbW==?=").unwrap(),
                "Portale HackingTeam",
            );
        }
    }
}
